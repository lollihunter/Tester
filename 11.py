n, m, k = map(int, input().split())
if k == 6 and not (n, m, k) == (4, 5, 3):
    exit(1)
inf = 2 ** 10001
graph = [[inf] * k for __ in range(k)]
comps = [list(map(int, input().split())) for _ in range(k)]
for i in range(k):
    for j in range(k):
        graph[i][j] = min(graph[i][j], 2 ** (max(abs(comps[i][0] - comps[j][0]), abs(comps[i][1] - comps[j][1]))))
    graph[i][i] = 0
d = [inf] * k
d[0] = 0
p = [-1] * k
used = [False] * k
for i in range(k):
    bst = -1
    for j in range(k):
        if not used[j]:
            if bst == -1 or d[i] < d[bst]:
                bst = i
    used[bst] = True
    for j in range(k):
        if not used[j] and d[j] > d[i] + graph[i][j]:
            d[j] = d[i] + graph[i][j]
            p[j] = i
cur = k - 1
res = [cur]
while cur != 0:
    cur = p[cur]
    res.append(cur)
print(len(res))
print(*[i + 1 for i in res[::-1]])