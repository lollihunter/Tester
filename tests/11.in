20 20 20
1 1
3 11
17 10
16 18
15 18
6 14
20 7
13 6
18 10
1 6
15 17
3 18
17 11
19 14
9 13
2 5
2 13
11 14
13 16
20 20
