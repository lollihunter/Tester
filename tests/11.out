11
1 16 10 2 6 15 18 19 11 4 20 
